Added select icon prop to replace the Select icon on the fly.

alansouzati opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
changes Select to add a new prop called `icon`.

#### Where should the reviewer start?
Select.js
#### What testing has been done on this PR?
manual
#### How should this be manually tested?
open storybook Select examples and make sure nothing breaks. also make sure the new ones allow for a custom icon or no icon.
#### Any background context you want to provide?

#### What are the relevant issues?
Closes #2935 
#### Screenshots (if appropriate)

#### Do the grommet docs need to be updated?
yes
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
backwards

ShimiSun commented on src/js/components/Select/doc.js:
```diff
@@ -77,6 +77,9 @@ export const doc = Select => {
     ),
+    icon: PropTypes.oneOfType([PropTypes.bool, PropTypes.node]).description(
+      'The icon to be used when rendering the select. You can use false to not render an icon at all.',
+    ),
```
I think it is missing `.defaultValue(true)`

alansouzati replied:
actually the default value is undefined.

ShimiSun replied:
We need to communicate that the icon is shown by default

alansouzati replied:
is that really needed? we dont communicate in other scenarios. take the `focusIndicator` right there... we don't say "we show the focus indicator by default".

alansouzati replied:
how about this for the description?

> A custom icon to be used when rendering the select. You can use false to not render an icon at all.

ShimiSun replied:
That should work, thanks

alansouzati commented:
I think we already have tests for `icon === undefined`, it is the default state.

oorestisime commented:
mmmm icon was added by this PR, i don't think there is a test for this. unless i miss something

alansouzati commented:
my point is `<Select />` is the same as `<Select icon={undefined} />`, no ?

oorestisime commented:
oh yeah sure. maybe just a test for icon false is all we need

ShimiSun commented:
We should have a tests for icon false and icon={}

alansouzati commented:
can `icon={true}` as well. i'm adding three tests.

oorestisime commented:
icon true should probably not be allowed? i think it breaks with current code at least. and not sure i see the point of `icon={}` value should be a node or a bool (or probably just false).

alansouzati commented:
i don't think it is possible to forbid the "icon={true}" if we want to allow `icon={false}`. I've fixed the code to account for the true scenario. it should never be used, but as you pointed out, it should not break the code. I've added a test for that.

alansouzati commented:
I've updated the PR with the new description for the icon prop. I think this should be good to merge now.

oorestisime commented:
haven't looked at the code but the caret down is not showing on storybook

oorestisime commented on src/js/components/Select/Select.js:
```diff
@@ -101,7 +102,10 @@ class Select extends Component {
-    const SelectIcon = theme.select.icons.down;
+    let SelectIcon;
+    if (icon !== false) {
+      SelectIcon = icon !== true ? icon : theme.select.icons.down;
```
breaks default state of icon undefined..
maybe a switch would be easier to read?
```
let SelectIcon;
switch (icon) {
  case false:
    break;
  case true:
    SelectIcon = theme.select.icons.down;
    break;
  case undefined:
    SelectIcon = theme.select.icons.down;
    break;
  default:
    SelectIcon = icon;
}
```

alansouzati replied:
great suggestion. let me work on it.

ShimiSun commented on src/js/components/Select/__tests__/Select-test.js:
```diff
@@ -356,4 +357,25 @@ describe('Select', () => {
+
+  test('renders custom icon', () => {
+    const component = renderer.create(
+      <Select id="test-select" options={['one', 'two']} icon={CaretDown} />,
```
PropTypes is not liking this one 
```
    ✓ renders custom icon (12ms)

 › 1 snapshot obsolete.
   • Select renders default icon 1
  console.error node_modules/react/node_modules/prop-types/checkPropTypes.js:19
    Warning: Failed prop type: Invalid prop `icon` supplied to `Select`.
        in Select (created by Select)
        in Select (created by Context.Consumer)
        in WithTheme(Select)
```

oorestisime replied:
So i had a look and since we are passing the component as `icon={CarretDown}` then this is not a react node but a function. i wonder though what would happen when someone passes an icon which is a class component.

Anyway proposed solution is to use `PropTypes.func` and also change the TS definition. This problem is also present in the Menu component.

alansouzati replied:
let me change it.
